### ISSUE TITLE
unable to load save games from Settlers 1 DOS

### ISSUE THREAD

**Issue by tlongstretch** [freeserf/freeserf#424]
I am attemping to load a savegame from a Settlers 1 game that has enemy AI players, with the intention of seeing if I can attack them to test out the combat code in FreeSerf.   I have no experience with C++ or Visual Studio so I am learning as I go.   

It seems I am unable to load any savegames from the original Settlers 1 game.  I am running "FreeSerf.exe -d 0 -l SAVE8.DS" on windows 7 from the command line.   The game window opens, music is heard, but an exception is thrown immediately before any graphics are seen.  I added some debug flags to the code to see where the error is thrown an the last place my debug statement showed was at the start of the savegame.cc SaveReader try/catch block.   I was able to kick off Visual Studio debugger which indicates the problem is with the ConfigFile::trim(std::string str) method's first line:  
> str.erase(str.begin(),

 The value of this->_vfptr ConfigFile type is:
> 0x014f0f38 {FreeSerf.exe!void(* ConfigFile::`vftable'[2])()} {0x012b8f94 {FreeSerf.exe!ConfigFile::`vector deleting destructor'(unsigned int)}}

I have attached screenshot of the error that pops up, along with some savegame files I am trying.  None work for me, all same error.

**Comment by tlongstretch** [comment]
![freeserf_load_game_exception](https://user-images.githubusercontent.com/45954803/50579299-ae33cb00-0e3b-11e9-87b1-bf7b0ec1d423.PNG)








[Settlers1_savegame_files.zip](https://github.com/freeserf/freeserf/files/2719846/Settlers1_savegame_files.zip)

**Comment by tlongstretch** [comment]
[Settlers1_savegame_files.zip](https://github.com/freeserf/freeserf/files/2719846/Settlers1_savegame_files.zip)

**Comment by tlongstretch** [comment]
![visual_studio_error](https://user-images.githubusercontent.com/45954803/50579323-f18e3980-0e3b-11e9-8d5f-75a0db4657f3.PNG)

**Comment by tlongstretch** [comment]
I now see that FreeSerf writes text savegame files which are easy to understand.  I will attempt to use this to  test the combat system by manually defining enemy player buildings in a savegame file and loading.

**Comment by tlongstretch** [comment]
Actually, it seems I am unable to read any save games at all, even ones created inside FreeSerf.  Here is the console output:

Debug: [savegame] inside GameStore::get_instance
Verbose: [viewport] map: 2048,1280, cols,rows: 64,64, tc,tr: 0,0, tw,th: 512,320
Verbose: [viewport] map: 2048,1280, cols,rows: 64,64, tc,tr: 1,0, tw,th: 512,320
Verbose: [viewport] map: 2048,1280, cols,rows: 64,64, tc,tr: 0,1, tw,th: 512,320
Verbose: [viewport] map: 2048,1280, cols,rows: 64,64, tc,tr: 1,1, tw,th: 512,320
Debug: [savegame] inside GameStore::get_instance
Debug: [savegame] inside GameStore::load
Debug: [savegame] passed file.is_open test
Debug: [savegame] inside SaveReader try/catch
Debug: [savegame] inside SaveReader try/catch after SaveReaderTextFile reader_text(&file) line
Warning: [savegame] Unable to load save game: ÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝ
Warning: [savegame] Trying compatability mode...
Error: [savegame] Failed to load save game: ÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝ®õ1ÝÝ
Debug: [savegame] inside GameStore::get_instance
Debug: [savegame] inside GameStore::load
Debug: [savegame] passed file.is_open test
Debug: [savegame] inside SaveReader try/catch
Debug: [savegame] inside SaveReader try/catch after SaveReaderTextFile reader_text(&file) line
Warning: [savegame] Unable to load save game: ÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÒü6
Warning: [savegame] Trying compatability mode...
Error: [savegame] Failed to load save game: ÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝƒ¯õ1
Debug: [game-manager] entering load_game
Debug: [game-manager] created new_game
Debug: [savegame] inside GameStore::get_instance
Debug: [savegame] inside GameStore::load
Debug: [savegame] passed file.is_open test
Debug: [savegame] inside SaveReader try/catch
Debug: [savegame] inside SaveReader try/catch after SaveReaderTextFile reader_text(&file) line
Warning: [savegame] Unable to load save game: ÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÒü6
Warning: [savegame] Trying compatability mode...
Error: [savegame] Failed to load save game: ÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝÝ¯õ1ÝÝ
Debug: [game-manager] failed to create new_game

**Comment by wdigger** [comment]
Fixed in release-0.3 branch.

**Comment by tlongstretch** [comment]
With the most recent checkout of master branch I can now load my DOS Settlers savegames.  However, once loaded I end up with the same issue described in #425 - that is, the interface banner does not load and any selections result in a game crash with error below
![freeserf_issue_425](https://user-images.githubusercontent.com/45954803/50728156-8daa9000-111d-11e9-9b2e-10b418a11d62.PNG)
![freeserf_issue_425_exception](https://user-images.githubusercontent.com/45954803/50728158-913e1700-111d-11e9-8743-b4779ceaf986.PNG)

**Comment by tlongstretch** [comment]
[SAVE8.zip](https://github.com/freeserf/freeserf/files/2729808/SAVE8.zip)

here is the saved game I am loading in this case

**Comment by tlongstretch** [comment]
exception traced in VS debugger  shows:

  Game *get_game() const { return game; }
  unsigned int get_index() const { return index; }   <----- here
![freeserf_issue_425_exception_trace](https://user-images.githubusercontent.com/45954803/50728207-1de8d500-111e-11e9-92de-cf47649ef283.PNG)

**Comment by wdigger** [comment]
I fix much bugs related to legacy game loading, please try to play now.

**Comment by tlongstretch** [comment]
Thank you!  I will try soon.  I am working on AI at the moment

**Comment by tlongstretch** [comment]
Did you apply fix to master branch?  That is where I am always pulling from.  I just now did a new build from master.

I still see the same issue with both legacy DOS and FreeSerf saves.  Both load properly but the interface banner does not appear and the game crashes upon clicking around.

**Comment by tlongstretch** [comment]
``b@PC /cygdrive/c/foo13/freeserf/src
$ diff /cygdrive/c/foo12/freeserf/src/savegame.h /cygdrive/c/foo13/freeserf/src/savegame.h
110a111
>   virtual bool has_value(const std::string &name) = 0;

b@-PC /cygdrive/c/foo13/freeserf/src
$ ^C

b@-PC /cygdrive/c/foo13/freeserf/src
$ diff /cygdrive/c/foo12/freeserf/src/savegame.cc /cygdrive/c/foo13/freeserf/src/savegame.cc
177a178,184
>
>   virtual bool has_value(const std::string &name) {
>     std::string v_name = name;
>     std::transform(v_name.begin(), v_name.end(), v_name.begin(), ::tolower);
>     Values::const_iterator it = values.find(v_name);
>     return (it != values.end());
>   }
226c233,235
<     Values::const_iterator it = values.find(name);
---
>     std::string v_name = name;
>     std::transform(v_name.begin(), v_name.end(), v_name.begin(), ::tolower);
>     Values::const_iterator it = values.find(v_name);
245a255,261
>   }
>
>   virtual bool has_value(const std::string &name) {
>     std::string v_name = name;
>     std::transform(v_name.begin(), v_name.end(), v_name.begin(), ::tolower);
>     Values::const_iterator it = values.find(v_name);
>     return (it != values.end());

**Comment by wdigger** [comment]
Which save do you mean?

**Comment by wdigger** [comment]
2 more cases fixed. I can load and play all your games.

**Comment by wdigger** [comment]
All fixes merged to master.

**Comment by tlongstretch** [comment]
Thank you for your amazing responsiveness!

When I load FreeSerf saves they seem to work properly, the interface banner loads.
However when I load a legacy DOS savegame the interface banner still does not load, and the game still crashes when clicking around.  I have attached the savegame I am testing with - SAVE1.DS

**Comment by wdigger** [comment]
I'm working on macOS, loads and works fine, maybe it's a some kind on platform specific. I'll try it with windows.

**Comment by tlongstretch** [comment]
I will try some other savegames

**Comment by sternenzaehler** [comment]
hi, just to add my 2 cents to this issue. the linux version can not properly load its own saved game (text format). After the loading the GUI panel is not showing (similar to the above description by tlongstretch) and my own player seems like an alien (can not click on buildings - crashes).
Repeat - this is NOT the legacy DOS save, just a normal freeserf save.
I built the linux version on ubuntu 16.04 x64 with the latest master and SDL 2.04.
The game itself seemed fine, I played for 3 hours on the 10ner map and saved. But was not able to load the game.

If you need some additional info I will be happy to help, but as it seems the bug is reproducible on linux by just trying to load any saved freeserf game.
Haven't tried any other platforms.

**Comment by wdigger** [comment]
@sternenzaehler please, attach save file here.

**Comment by wdigger** [comment]
@tlongstretch I try SAVE1.DS from Settlers1_savegame_files.zip archive on macOS and Windows, works fine for me. Is it file are you saying about?

**Comment by tlongstretch** [comment]
@widgger
Yes that is the one.  I will try it again (and some others) soon with a clean checkout from git.   Right now my recent checkout is polluted with my own AI code that I have been testing so it isn't a reliable test.

**Comment by sternenzaehler** [comment]
[savefreeserf.txt](https://github.com/freeserf/freeserf/files/2771165/savefreeserf.txt)
here is my freeserf saved game. This game loads fine but does not allow to play and crashes on any attempt of action (like try to click on any building)

by the way, here is one binary legacy savegame that leads to a crash right after loading. This is a complicated game on an 8er map with four players and many battles going on between them, the amount of serfs is over 16000. This game loads, starts to play and crashes in less than a second.

all tests on ubuntu 16.04 x64 using the Latest commit 6598afc

**Comment by sternenzaehler** [comment]
[SAVE0.DS.zip](https://github.com/freeserf/freeserf/files/2771181/SAVE0.DS.zip)
this is the "complicated" savegame. Not sure if it is time to test it now because I am not sure that freeserf is able to simulate four players with battles and free wandering knights engaging each other.
I was not expecting this to work, just tried.

By the way, is there any tool with freeserf that can convert native binary saves to freeserf text saves ?
maybe you guys have written one for developing purposes??

**Comment by wdigger** [comment]
@sternenzaehler 
> maybe you guys have written one for developing purposes??
yes, it's possible, but I can't see any reason, FreeSerf must load legacy saves as fine. You may save your game in new format after load, if you wish.
In any case, please, file new issue about this. I mark it as enchantment and maybe somehow...

**Comment by wdigger** [comment]
> [SAVE0.DS.zip](https://github.com/freeserf/freeserf/files/2771181/SAVE0.DS.zip)
> this is the "complicated" savegame. Not sure if it is time to test it now because I am not sure that freeserf is able to simulate four players with battles and free wandering knights engaging each other.

I fixed you game loading, please check it with last build from master.

**Comment by sternenzaehler** [comment]
> @sternenzaehler
> your textually saved game I can successfully load and play, may be someone who take a crash debug it by himself...

Hi! But have you tried it on linux? (as stated above, this might be platform specific).
Anyway, point taken, I will debug it by myself. By the way the crash is not the primary bug. The primary bug is that the GUI does not show (it somehow indicates that loading the savegame did not finish as intended despite the fact that the game plays).

As for the native binary SAVE0.DS your fix worked, it now loads and plays, but with the same bug that any click leads to a crash.

I will investigate this myself for linux.

By the way, I got the following errors in the log, is it something serious or is this expected?
Error: [data] Failed to get animation #128 phase #48 (got only 48 phases)
AND
Error: [data] Failed to get animation #126 phase #38 (got only 38 phases)

**Comment by sternenzaehler** [comment]
I seem to have found some strange behavior in the loading game sequence and was able to fix the bug for me. But I think I would need your help because It is only my first day with the code :)
The problem is the following.
Interface::set_player is supposed to be called from the on_new_game handler which is invoked by GameManager::set_current_game (which is in turn called from GameManager::load_game).
But when we load a game with the -l option here in freeserf.cc (line 113)
```
/* Either load a save game if specified or
     start a new game. */
  if (!save_file.empty()) {
    if (!game_manager.load_game(save_file)) {
      return EXIT_FAILURE;
  
```
there is yet no Interface instance and no handlers. The Interface gets instantiated after that 
freeserf.cc line 123
`/* Initialize interface */
  Interface interface;
  if ((screen_width == 0) || (screen_height == 0)) {
    gfx.get_resolution(&screen_width, &screen_height);
  }
  interface.set_size(screen_width, screen_height);
  interface.set_displayed(true);
`

so the on_new_game was not called because there was no instance.
And the Interface constructor does not call set_player either, it only does this 
interface.cc line 79
`  GameManager::get_instance().add_handler(this);
  set_game(GameManager::get_instance().get_current_game());
`
compare to on_new_game
`void
Interface::on_new_game(PGame new_game) {
  
  set_game(new_game);
  set_player(0);
}`

I tried set_player(0) in the Interface constructor and it works for me.
The reason I write this here instead of fixing it with a pull request (two of them)
1. You stated it works for you, and that is very strange because without set_player a loaded game wont work. it will play but you cannot control it and it will crash on any action attempt. for example the J command will result in an endless loop because current player would be rubbish. So maybe this is more complicated and it is only a bug on my linux system? The must be some reason why it works for you and not for me.
2. This bug concerns the initial loading and instantiation sequence which is very important for the game setup and can have consequences - I'd better not touch for now - it is best fixed by you as the code owner. I cannot say that set_player(0) in the constructor of Interface is a good idea or which consequences it might have.

**Comment by wdigger** [comment]
Usually I load game from UI, the key detail here - loading with command line ;)

**Comment by sternenzaehler** [comment]
Hi! Thank you very much for the fix. this is indeed better to call set_player directly from set_game than my solution .

Loading a game from UI on Linux is not possible(from my point of view), it just shows an empty black rectangle (which is supposed to be a list of saved games) and does not do anything.
By the way, it is not that easy to find a saved game in linux. 
it places the savegame into ~/.local/share/freeserf/saves/.save  
where nobody is gonna find it. I did not even know that it existed and thought it just does not work, until I saw this path hardcoded in the source code:)

I will try to load with the UI once more and try to look into it in more detail.

**Comment by sternenzaehler** [comment]
And one more thing, the GUI is very unresponsive (lagging up to 20 seconds after the click) if the number of serfs is big enough(more than 10 000). Is that a known problem or some issue on my part?
CPU load is 100% of a core when playing such a game (example can be found above - the binary SAVE0.DS from my earlier comment).
The game itself is not slow - it is normally paced, but the GUI becomes really slow.

**Comment by wdigger** [comment]
@sternenzaehler 
please, file issues for slow and laggy UI and bad animation warning, it's better to have one ticket for each problem.
Do you have any suggestions about save files placement on Linux?

**Comment by sternenzaehler** [comment]
Hi, 
I think it is  always better (for open source projects) to ask first if it is a known problem and not important than to just blindly file tickets for all I see. 

That is why I asked you whether the anim. warning was important for now or can be ignored.
The Issue with lagging GUI is an important one in comparison.

But if you would like it, I can file many tickets then :)

**Comment by sternenzaehler** [comment]
Suggestion for savegames on linux: 
place them to ~/.freeserf, not ~/.local/share/freeserf. 
this is more or less an expected behaviour, most programs keep their data like this.
But this is my opinion and each linux user has a different opinion about this :)

I think it is more important to fix the GUI loading. If the game can load with the GUI then it is not that important where savegames are located. I only used command line because the GUI does not work on a linux system. I will file an issue about this as well as soon as I have inspected the details.
